pkg/plugins/storage/azblob/azblob.go

```diff
@@ -0,0 +1,185 @@
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/nitric-dev/membrane/pkg/plugins/errors"
+	"github.com/nitric-dev/membrane/pkg/plugins/errors/codes"
+	"github.com/nitric-dev/membrane/pkg/plugins/storage"
+	azblob_service_iface "github.com/nitric-dev/membrane/pkg/plugins/storage/azblob/iface"
+	azureutils "github.com/nitric-dev/membrane/pkg/providers/azure/utils"
+	"github.com/nitric-dev/membrane/pkg/utils"
+)
+
+// AzblobStorageService - Nitric membrane storage plugin implementation for Azure Storage
+type AzblobStorageService struct {
+	client azblob_service_iface.AzblobServiceUrlIface
+}
+
+func (a *AzblobStorageService) getBlobUrl(bucket string, key string) azblob_service_iface.AzblobBlockBlobUrlIface {
+	cUrl := a.client.NewContainerURL(bucket)
+	// Get a new blob for the key name
+	return cUrl.NewBlockBlobURL(key)
+}
+
+func (a *AzblobStorageService) Read(bucket string, key string) ([]byte, error) {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Read",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+	// Get the bucket for this bucket name
+	blob := a.getBlobUrl(bucket, key)
+	//// download the blob
+	r, err := blob.Download(
+		context.TODO(),
+		0,
+		azblob.CountToEnd,
+		azblob.BlobAccessConditions{},
+		false,
+		azblob.ClientProvidedKeyOptions{},
+	)
+
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"Unable to download blob",
+			err,
+		)
+	}
+
+	// TODO: Configure retries
+	data := r.Body(azblob.RetryReaderOptions{MaxRetryRequests: 20})
+
+	return ioutil.ReadAll(data)
+}
+
+func (a *AzblobStorageService) Write(bucket string, key string, object []byte) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Write",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Upload(
+		context.TODO(),
+		bytes.NewReader(object),
+		azblob.BlobHTTPHeaders{},
+		azblob.Metadata{},
+		azblob.BlobAccessConditions{},
+		azblob.DefaultAccessTier,
+		nil,
+		azblob.ClientProvidedKeyOptions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to write blob data",
+			err,
+		)
+	}
+
+	return nil
+}
+
+func (a *AzblobStorageService) Delete(bucket string, key string) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Delete",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	// Get the bucket for this bucket name
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Delete(
+		context.TODO(),
+		azblob.DeleteSnapshotsOptionInclude,
+		azblob.BlobAccessConditions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to delete blob",
+			err,
+		)
+	}
+
+	return nil
+}
+
+const expiryBuffer = 2 * time.Minute
```

Comment:
Assume this doesn't require config? 2 mins is always adequate for the buffer?

Reply:
For now, can address if it becomes a problem in future.